public/app/features/explore/Table/TableContainer.tsx

```diff
@@ -85,33 +83,45 @@ export class TableContainer extends PureComponent<Props> {
       }
     }
 
-    const mainFrame = this.getMainFrame(dataFrames);
-    const subFrames = dataFrames?.filter((df) => df.meta?.custom?.parentRowIndex !== undefined);
+    const tableData: Array<{ main: DataFrame; sub?: DataFrame[] }> = [];
+    const mainFrames = this.getMainFrames(dataFrames).filter(
+      (frame: DataFrame | undefined): frame is DataFrame => !!frame && frame.length !== 0
+    );
+
+    mainFrames?.forEach((frame) => {
+      const subFrames =
+        dataFrames?.filter((df) => frame.refId === df.refId && df.meta?.custom?.parentRowIndex !== undefined) ||
+        undefined;
+      tableData.push({ main: frame, sub: subFrames });
+    });
```

Comment:
tbh i find this logic is becoming a bit confusing and starting to become a bit too spread over this component.
wdyt about this:

```ts
    const tableData = dataFrames?.filter(isMainFrame).map((mainFrame) => ({
      main: mainFrame,
      sub: dataFrames?.filter(isChildOf(mainFrame)),
    }));
```

where `isMainFrame` and `isChildOf` are defined as (outside of the component):

```ts
function isMainFrame(frame: DataFrame) {
  return frame.meta?.custom?.parentRowIndex === undefined;
}

function isChildOf(parentFrame: DataFrame) {
  return (maybeChild: DataFrame) => {
    return parentFrame.refId === maybeChild.refId && maybeChild.meta?.custom?.parentRowIndex !== undefined;
  };
}
```

Reply:
The whole splitting logic could be encapsulated in a single function. Maybe not in this PR, but do you think it would make sense to make it part of the decorator logic?

Reply:
i don't see why not, it would actually make more sense there than here, especially given the recent progression over custom viz which may benefit from it as well